### PR TITLE
feat(gatsby): page engine

### PR DESCRIPTION
**Pull request by pieh**
I suggest to use "Hide whitespace changes" in diff viewer - https://github.com/gatsbyjs/gatsby/pull/33067/files?diff=unified&w=1

Building on top of GraphQL Engine introduced in https://github.com/gatsbyjs/gatsby/pull/33030

## Description

> This feature is disabled by default. Only available when building gatsby itself with `COMPILER_OPTIONS="GATSBY_MAJOR=4"`

This adds generation of "page engine" - importable standalone module that allows generation of page-data and html content. It uses "query engine" (introduced in https://github.com/gatsbyjs/gatsby/pull/33030 ) under the hood.

Example usage (after initial `gatsby build`):

```js
const { GraphQLEngine } = require(`./.cache/query-engine`)
const { getData, renderHTML, renderPageData } = require(`./.cache/page-ssr`)

const graphqlEngine = new GraphQLEngine({
  dbPath: process.cwd() + `/.cache/data/datastore`,
})

async function run() {
  const data = await getData({ graphqlEngine, pathName })
  const pageData = await renderPageData({ data })
  const html = await renderHTML({ data, pageData })

  console.log({ pageData, html })
}

run()
``` 

[ch37670]

Fixes https://github.com/gatsbyjs/gatsby/issues/32892

### CODE REVIEW

File: packages/gatsby/src/commands/build.ts [hunk 1]
```diff
@@ -232,14 +238,34 @@ module.exports = async function build(program: IBuildArgs): Promise<void> {
   try {
     const result = await buildRenderer(program, Stage.BuildHTML, buildSpan)
     pageRenderer = result.rendererPath
+    if (_CFLAGS_.GATSBY_MAJOR === `4` && shouldGenerateEngines()) {
+      // for now copy page-render to `.cache` so page-ssr module can require it as a sibling module
+      const outputDir = path.join(program.directory, `.cache`, `page-ssr`)
```
Comment (wardpeet) [inline packages/gatsby/src/commands/build.ts:243]: Can you rename it to _routes as it's more in par with https://github.com/gatsbyjs/gatsby/pull/33054
```suggestion
      const outputDir = path.join(program.directory, `.cache`, `_routes`)
```

File: packages/gatsby/src/utils/client-assets-for-template.ts [hunk 2]
```diff
@@ -0,0 +1,153 @@
+import * as path from "path"
```
Comment (pieh) [inline packages/gatsby/src/utils/client-assets-for-template.ts:1]: Extracted to separate module from `packages/gatsby/src/utils/worker/child/render-html.ts`

File: packages/gatsby/src/utils/page-data-helpers.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
+import { IGatsbyPage } from "../redux/types"
```
Comment (pieh) [inline packages/gatsby/src/utils/page-data-helpers.ts:1]: Extracted to separate module from `packages/gatsby/src/utils/page-data.ts` + addition of `getPagePathFromPageDataPath` to reverse `/page-data/<page-path>/page-data.json` to `<page-path>`

File: packages/gatsby/src/utils/static-query-utils.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
+import fs from "fs-extra"
```
Comment (pieh) [inline packages/gatsby/src/utils/static-query-utils.ts:1]: Extracted to separate module from `packages/gatsby/src/utils/worker/child/render-html.ts`

File: packages/gatsby/src/commands/build.ts [hunk 5]
```diff
@@ -232,14 +238,34 @@ module.exports = async function build(program: IBuildArgs): Promise<void> {
...
+    }
     waitForCompilerCloseBuildHtml = result.waitForCompilerClose
   } catch (err) {
     buildActivityTimer.panic(structureWebpackErrors(Stage.BuildHTML, err))
   } finally {
     buildSSRBundleActivityProgress.end()
   }
 
+  if (_CFLAGS_.GATSBY_MAJOR === `4` && shouldGenerateEngines()) {
+    // well, tbf we should just generate this in `.cache` and avoid deleting it :shrug:
```
Comment (vladar) [inline packages/gatsby/src/commands/build.ts:263]: 👍 I assume doing this in v3 will be a breaking change?

File: packages/gatsby/src/utils/client-assets-for-template.ts (hunk 2 above)
Comment (vladar) [reply packages/gatsby/src/utils/client-assets-for-template.ts:1]: Just to clarify - we need to have it as a separate module because we re-use it later in bundle-webpack, right?

File: packages/gatsby/src/commands/build.ts (hunk 5 above)
Comment (pieh) [reply packages/gatsby/src/commands/build.ts:263]: Possibly yeah, I just didn't want to introduce more changes in this PR than absolutely needed.

File: packages/gatsby/src/utils/page-data.ts [hunk 6]
```diff
@@ -119,36 +114,21 @@ export async function readPageQueryResult(
...
-    staticQueryHashes,
-  }: IPageData
+  pageData: IPageData
 ): Promise<string> {
-  const result = await readPageQueryResult(publicDir, pagePath)
-
-  const outputFilePath = generatePageDataPath(publicDir, pagePath)
-  let body = `{
-    "componentChunkName": "${componentChunkName}",
-    "path": "${pagePath}",
```
Comment (vladar) [inline packages/gatsby/src/utils/page-data.ts:134]: _offtopic for this PR_:  quotes are not escaped here. I wonder if it is the cause of https://github.com/gatsbyjs/gatsby/issues/32892

File: packages/gatsby/src/utils/page-ssr-module/bundle-webpack.ts [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        {
+          test: /\.txt/,
+          type: `asset/resource`,
+        },
+      ],
+    },
+    resolve: {
+      extensions,
+      alias: {
+        ".cache": process.cwd() + `/.cache/`,
```
Comment (vladar) [inline packages/gatsby/src/utils/page-ssr-module/bundle-webpack.ts:103]:
```suggestion
        ".cache": `${program.directory}/.cache/`,
```

File: packages/gatsby/src/utils/page-ssr-module/entry.ts [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    {
+      componentChunkName: data.page.componentChunkName,
+      path: data.page.path,
+      matchPath: data.page.matchPath,
+      staticQueryHashes: data.templateDetails.staticQueryHashes,
+    },
+    JSON.stringify(data.results)
+  )
+
+  return JSON.parse(results)
```
Comment (vladar) [inline packages/gatsby/src/utils/page-ssr-module/entry.ts:91]: I assume we have to parse here because of #32763 

Looks a bit weird and we should probably double-check at some point if the benefits of that PR outweigh the overhead here.

Comment (pieh) [reply packages/gatsby/src/utils/page-ssr-module/entry.ts:91]: I think it's possible that we can/will drop `JSON.parse` here and return stringified version of `page-data` instead of json blob. Just I wouldn't want to make those changes in this PR because it's not quite clear to me if it would be safe to do so or not

**Review by vladar — APPROVED** [review]
LGTM 👍

File: packages/gatsby/src/utils/page-data-helpers.ts [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    componentChunkName,
+    matchPath,
+    path: pagePath,
+    staticQueryHashes,
+  }: IPageData,
+  result: string | Buffer
+): string {
+  let body = `{
+    "componentChunkName": "${componentChunkName}",
+    "path": "${pagePath}",
```
Comment (vladar) [inline packages/gatsby/src/utils/page-data-helpers.ts:21]: I think we can safely change it along the way to fix https://github.com/gatsbyjs/gatsby/issues/32892. But not insisting - could be a separate PR as well.
```suggestion
    "path": JSON.stringify(pagePath),
```

Comment (pieh) [reply packages/gatsby/src/utils/page-data-helpers.ts]: Makes sense, comitted that suggestion in

File: packages/gatsby/src/utils/page-data.ts (hunk 6 above)
Comment (pieh) [reply packages/gatsby/src/utils/page-data.ts:134]: I think with https://github.com/gatsbyjs/gatsby/pull/33067/commits/90a2c57faa70e7fe5420d44237e6245b68cc1985 we can resolve this one?

**Review by vladar — APPROVED** [review]

File: packages/gatsby/src/utils/page-data.ts (hunk 6 above)
Comment (vladar) [reply packages/gatsby/src/utils/page-data.ts:134]: Yup 👍

File: packages/gatsby/src/utils/client-assets-for-template.ts (hunk 2 above)
Comment (pieh) [reply packages/gatsby/src/utils/client-assets-for-template.ts:1]: Hmm, possibly we could revert this split because we don't use it in engine itself. But it's still beneficial because we don't have to load more things from workers code than needed (`child/render-html.ts` is not loaded by main process, just by html/pqr worker processes)

File: packages/gatsby/src/utils/page-data-helpers.ts [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    componentChunkName,
+    matchPath,
+    path: pagePath,
+    staticQueryHashes,
+  }: IPageData,
+  result: string | Buffer
+): string {
+  let body = `{
+    "componentChunkName": "${componentChunkName}",
+    "path": JSON.stringify(pagePath),
```
Comment (vladar) [inline packages/gatsby/src/utils/page-data-helpers.ts:21]: I've messed it up :)

```suggestion
    "path": ${JSON.stringify(pagePath)},
```

Comment (pieh) [reply packages/gatsby/src/utils/page-data-helpers.ts:21]: lol, I didn't catch that either :)

**Review by vladar — APPROVED** [review]
